haystack/preview/components/reader.py

```diff
@@ -0,0 +1,255 @@
+        not_last_docs = []
+        i = 0
+        for j in range(query_ids[-1]):
+            while i < len(query_ids) and query_ids[i] == j:
+                if i != 0:
+                    not_last_docs.append(i - 1)
+                i += 1
+            last_docs.append(i - 1)
+
+        # The following might seem a bit obfuscated, but the goal of doing it like this is to keep the operations vectorized and on the GPU
+        # Sum all no answer logits for the query and only save in the last slice
+        no_answer_logits = logits[..., 0, 0]
+        no_answer_logits_sum = torch.cumsum(no_answer_logits, -1)
+        no_answer_logits_sum[last_docs[1:]] -= no_answer_logits_sum[last_docs[:-1]]
+        no_answer_logits_sum[not_last_docs] = -torch.inf
+        logits[..., 0, 0] = no_answer_logits_sum
+
+        exp_logits = torch.exp(logits)
+        # For softmax, we need to sum all exp_logits for each query to normalise them
+        # This is a bit more complicated because there can be multiple queries in one batch
+        # There is also a varying number of slices and documents per query so we can't just create a new axis
+        # Start by computing sum for every slice across start and end axis
+        exp_logits_sum = torch.sum(exp_logits, -1)  # Sum across end axis
+        exp_logits_sum = torch.sum(exp_logits_sum, -1)  # Sum across start axis
+
+        # Sum across all samples
+        exp_logits_sum = torch.cumsum(exp_logits_sum, -1)
+        # Every last sample of a query will have the sum of all samples belonging to the query + all other samples belonging to other queries
+        # By subtracting the value of the previous last sample, we get the sum of the items only belonging to the respective query
+        exp_logits_sum[not_last_docs] = 0
+        exp_logits_sum[last_docs[1:]] -= exp_logits_sum[last_docs[:-1]]
+        # We now have the correct sum at the indices containing the last sample for a query
+        # However, all other values are 0
+        # We can propagate these values to the other samples using cumsum again
+        # To do that, we need to revert the axis so the correct value is the first sample for each query as cumsum only goes from left to right
+        rev_sum = torch.flip(exp_logits_sum, [-1])
+        rev_cumsum = torch.cumsum(rev_sum, -1)
+        # We can again subtract the previous values so we only have the value for the current query
+        exp_logits_sum[last_docs[:-1]] = exp_logits_sum[last_docs[1:]]
+        exp_logits_sum[-1] = 0
+        rev_sum_shifted = torch.flip(exp_logits_sum, [-1])
+        rev_sum_shifted = torch.cumsum(rev_sum_shifted, -1)
+        rev_cumsum -= rev_sum_shifted
+        # Having used the reversed order for cumsum, we can revert the axis back
+        exp_sum_per_query = torch.flip(rev_cumsum, [-1])
+
+        # Now, we can finally use the sums to normalise the values
+        exp_logits /= exp_sum_per_query.unsqueeze(-1).unsqueeze(-1)
+        return exp_logits
```

Comment:
@Timoeller, @julian-risch or @bogdankostic might be our best bets

Reply:
I don't think we should go this way, it would be very unintuitive from an end-user perspective. Imagine having two Documents as input, one with multiple good answers and another one with a mediocre answer. If I understand your comment correctly, the good answers would get a lower score than the mediocre answer just because of the fact that they come from a Document containing many good answers. As an end-user, I would perceive this as the model not working well.

Reply:
Just to chip in here, the Sol team also found the way FARMReader calculated confidence scores to be unintuitive and somewhat confusing. Also, we found that the implementation in v1 artificially lowers the probability of correct answers if they happen to be clustered together, which can then cause a mediocre answer to be pushed to the top. 

In Sol, we actually discussed using the logits for each answer directly instead so we could better compare answers across documents. And we have tried this out for one of our clients which has worked well. 

So I like the idea of @MichelBartels implementation, but I agree that running a softmax across all answers may lower the probabilities overall (since they all need to add up to 1) which may be confusing to users. One idea that I had (but haven't tried) is to consider applying a `torch.nn.Sigmoid()` function directly to each of the logits to get a probability between 0 and 1 for each answer. This is how the `SentenceTransformersRanker` and I believe the `EntityExtractor` produces probabilities for their predictions. This should still preserve the ordering from @MichelBartels implementation, but might provide more intuitive confidence scores.

Reply:
Thanks for the feedback. I like the idea of using sigmoid and will try if it works. However, I think it will probably need some form of (perhaps even model-specific) temperature because the logits are often quite big. For example, in @ZanSara's test code the logit for Angela Merkel which is dominated by the no-answer logit is still around 12 which would translate to 99.999% confidence.

Reply:
I remember having this discussion many times as well back then when implementing it: The softmax across all logits seemed like a bad thing to do in theory. Though in practice it works refreshingly well. I would not spend effort changing this during the migration. If you do, you might need to change the aggregation logic (including no_answer aggregation).

Reply:
Sorry, we discussed this during the co:work event and I forgot to add our findings here. We had agreed that using sigmoid seems promising although you probably need some function to transform the logits, which we decided is just some constant factor for now.
We also realised that the no answer score customers are looking for is probably not the same as the score returned by the models (even without splitting problems). We agreed that the customer only cares about the probability that the correct answers is not in the top_k and not about the probability that the answer exists in the document (which SQuAD cares) about. If we assume that and take each span as a separate classification task, then no answer is just the product of the complementary probabilities of all the answer in the top k. I have implemented it this way in the PR, but I can also change it back.
@sjrl I would be interested in what you think customers want no answer scores to represent.
@Timoeller, I am not sure I understand your point about the aggregation logic. Did you mean that the way it was implemented in the PR needs to be changed? Or did you mean that it would need to be different from what `FARMReader` does (which it already was)?

Reply:
> We agreed that the customer only cares about the probability that the correct answers is not in the top_k and not about the probability that the answer exists in the document (which SQuAD cares about).

I would agree with this statement! To be honest we actually have turned off No Answer predictions completely for all customer use cases because of how unreliable and confusing they were. 

The only exception here is for one customer, but they only return No Answers because in their downstream application, they would still like to return all documents even if that individual document didn't have answers that contributed to the top_k. So even they weren't so interested in the No Answer prediction due to the same reason as I stated above, but rather would like to see all documents that were provided as input to the ExtractiveReader.